### PR TITLE
*: add admin cleanup table lock syntax support

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Related Parser PR: https://github.com/pingcap/parser/pull/323.

Add cleanup table lock syntax:
```sql
ADMIN CLEANUP TABLE LOCK table_name1 [, table_name2...]
```
Cleanup table lock use to cleanup table locks. Normally won't use it. Just in case for session has not clean locks before exit.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -3366,10 +3366,66 @@ func (d *ddl) UnlockTables(ctx sessionctx.Context, unlockTables []model.TableLoc
...
 }
 
+func (d *ddl) CleanupTableLock(ctx sessionctx.Context, tables []*ast.TableName) error {
+	uniqueTableID := make(map[int64]struct{})
+	cleanupTables := make([]model.TableLockTpInfo, 0, len(tables))
+	// Check whether the table was already locked by other.
+	for _, tb := range tables {
+		// TODO: replace const string "performance_schema" with xxx.LowerName.
+		// Currently use perfschema.LowerName will have import cycle problem.
+		if tb.Schema.L == infoschema.LowerName || tb.Schema.L == "performance_schema" || tb.Schema.L == mysql.SystemDB {
```
Comment (winkyao) [inline ddl/ddl_api.go]: I think you better to use a function for this check, I found you repeat the code in https://github.com/pingcap/tidb/blob/a935d93990ad7059d3f2a07eeec1eda71f257353/ddl/ddl_api.go#L3401

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -3366,10 +3366,66 @@ func (d *ddl) UnlockTables(ctx sessionctx.Context, unlockTables []model.TableLoc
...
+		return nil
+	}
+
+	arg := &lockTablesArg{
+		UnlockTables: cleanupTables,
+		IsCleanup:    true,
+	}
+	job := &model.Job{
+		SchemaID:   cleanupTables[0].SchemaID,
+		TableID:    cleanupTables[0].TableID,
```
Comment (winkyao) [inline ddl/ddl_api.go]: Just specify the first tableID? Does it work in ApplyDiff?

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -1978,6 +1978,10 @@ func (b *PlanBuilder) buildDDL(node ast.DDLNode) (Plan, error) {
 		b.visitInfo = appendVisitInfo(b.visitInfo, mysql.SuperPriv, "", "", "", nil)
 	case *ast.LockTablesStmt, *ast.UnlockTablesStmt:
 		// TODO: add Lock Table privilege check.
+	case *ast.CleanupTableLockStmt:
+		// This command can only be executed by administrator.
+		b.visitInfo = appendVisitInfo(b.visitInfo, mysql.SuperPriv, "", "", "", nil)
+
```
Comment (zimulala) [inline planner/core/planbuilder.go]: Please remove this useless line.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -3474,10 +3471,72 @@ func (d *ddl) UnlockTables(ctx sessionctx.Context, unlockTables []model.TableLoc
 	return errors.Trace(err)
 }
 
+func checkLockInMemOrSysDB(ctx sessionctx.Context, dbLowerName string) error {
+	if util.IsMemOrSysDB(dbLowerName) {
+		if ctx.GetSessionVars().User != nil {
+			return infoschema.ErrAccessDenied.GenWithStackByArgs(ctx.GetSessionVars().User.Username, ctx.GetSessionVars().User.Hostname)
+		}
+		return infoschema.ErrAccessDenied.GenWithStackByArgs()
```
Comment (zimulala) [inline ddl/ddl_api.go]: Do we need to add an argument for this function?

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -3474,10 +3471,72 @@ func (d *ddl) UnlockTables(ctx sessionctx.Context, unlockTables []model.TableLoc
...
+		err := checkLockInMemOrSysDB(ctx, tb.Schema.L)
+		if err != nil {
+			return err
+		}
+		schema, t, err := d.getSchemaAndTableByIdent(ctx, ast.Ident{Schema: tb.Schema, Name: tb.Name})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if t.Meta().IsView() {
+			return table.ErrUnsupportedOp.GenWithStackByArgs()
```
Comment (zimulala) [inline ddl/ddl_api.go]: Could we  remove `GenWithStackByArgs`?

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -3474,10 +3471,72 @@ func (d *ddl) UnlockTables(ctx sessionctx.Context, unlockTables []model.TableLoc
 	return errors.Trace(err)
 }
 
+func checkLockInMemOrSysDB(ctx sessionctx.Context, dbLowerName string) error {
```
Comment (tangenta) [inline ddl/ddl_api.go]: I think this function is nothing to do with `Lock`?
IMHO, it should name as `throwErrIfInMemOrSysDB`.

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -3474,10 +3471,72 @@ func (d *ddl) UnlockTables(ctx sessionctx.Context, unlockTables []model.TableLoc
...
+		schema, t, err := d.getSchemaAndTableByIdent(ctx, ast.Ident{Schema: tb.Schema, Name: tb.Name})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if t.Meta().IsView() {
+			return table.ErrUnsupportedOp.GenWithStackByArgs()
+		}
+		tbInfo := t.Meta()
+
+		if tbInfo.Lock == nil || len(tbInfo.Lock.Sessions) == 0 {
```
Comment (zimulala) [inline ddl/ddl_api.go]: If this DDL job is after the other DDL job that locks this table. Then we do this DDL job. Do we need to clean up this table's lock?

Comment (crazycs520) [reply ddl/ddl_api.go]: Great, Thanks.

**Comment by crazycs520** [comment]
@bb7133 PTAL

**Comment by AilinKid** [comment]
LGTM

**Pull request merged**